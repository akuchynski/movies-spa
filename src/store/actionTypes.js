export default {

    GET_MOVIES_REQUEST: 'GET_MOVIES_REQUEST',
    GET_MOVIES_SUCCESS: 'GET_MOVIES_SUCCESS',
    GET_MOVIES_ERROR: 'GET_MOVIES_ERROR',

    GET_MOVIE_REQUEST: 'GET_MOVIE_REQUEST',
    GET_MOVIE_SUCCESS: 'GET_MOVIE_SUCCESS',
    GET_MOVIE_ERROR: 'GET_MOVIE_ERROR',

    CLOSE_MOVIE_DETAILS: 'CLOSE_MOVIE_DETAILS',
    SET_ACTIVE_FILTER: 'SET_ACTIVE_FILTER',

    ADD_MOVIE_SUCCESS: 'ADD_MOVIE_SUCCESS',
    ADD_MOVIE_ERROR: 'ADD_MOVIE_ERROR',

    UPDATE_MOVIE_SUCCESS: 'UPDATE_MOVIE_SUCCESS',
    UPDATE_MOVIE_ERROR: 'UPDATE_MOVIE_ERROR',

    DELETE_MOVIE_SUCCESS: 'DELETE_MOVIE_SUCCESS',
    DELETE_MOVIE_ERROR: 'DELETE_MOVIE_ERROR',

};