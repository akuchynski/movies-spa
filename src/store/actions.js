import actionTypes from "./actionTypes";

const getMoviesRequest = () => ({
    type: actionTypes.GET_MOVIES_REQUEST,
});

const getMoviesSuccess = (movies) => ({
    type: actionTypes.GET_MOVIES_SUCCESS,
    payload: movies
});

const getMoviesError = (errorMessage) => ({
    type: actionTypes.GET_MOVIES_ERROR,
    payload: errorMessage
});

const getMovieRequest = () => ({
    type: actionTypes.GET_MOVIE_REQUEST,
});

const getMovieSuccess = (movie) => ({
    type: actionTypes.GET_MOVIE_SUCCESS,
    payload: movie
});

const getMovieError = (errorMessage) => ({
    type: actionTypes.GET_MOVIE_ERROR,
    payload: errorMessage
});

const closeMovieDetails = () => ({
    type: actionTypes.CLOSE_MOVIE_DETAILS,
});

const setActiveFilter = (value) => ({
    type: actionTypes.SET_ACTIVE_FILTER,
    payload: value
});

export default {
    getMoviesRequest,
    getMoviesSuccess,
    getMoviesError,
    getMovieRequest,
    getMovieSuccess,
    getMovieError,
    closeMovieDetails,
    setActiveFilter,
};