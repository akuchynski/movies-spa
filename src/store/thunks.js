import axios from "axios";
import actions from "./actions";

export const getMovies = () => (dispatch) => {
    dispatch(actions.getMoviesRequest());
    axios.get('http://localhost:4000/movies')
        .then((response) => dispatch(actions.getMoviesSuccess(response.data)))
        .catch((error) => dispatch(actions.getMoviesError(error.message)));
};

export const getMoviesByGenre = (genre) => (dispatch) => {
    dispatch(actions.getMoviesRequest());
    axios.get('http://localhost:4000/movies', genre !== 'ALL' && { params: { searchBy: 'genres', filter: genre } })
        .then((response) => dispatch(actions.getMoviesSuccess(response.data)))
        .catch((error) => dispatch(actions.getMoviesError(error.message)));
};

export const getMoviesByDate = () => (dispatch) => {
    dispatch(actions.getMoviesRequest());
    axios.get('http://localhost:4000/movies?sortBy=release_date&sortOrder=desc')
        .then((response) => dispatch(actions.getMoviesSuccess(response.data)))
        .catch((error) => dispatch(actions.getMoviesError(error.message)));
};

export const getMoviesByRating = () => (dispatch) => {
    dispatch(actions.getMoviesRequest());
    axios.get('http://localhost:4000/movies?sortBy=vote_average&sortOrder=desc')
        .then((response) => dispatch(actions.getMoviesSuccess(response.data)))
        .catch((error) => dispatch(actions.getMoviesError(error.message)));
};

export const getMovieById = (id) => (dispatch) => {
    dispatch(actions.getMovieRequest());
    axios.get('http://localhost:4000/movies/' + id)
        .then((response) => dispatch(actions.getMovieSuccess(response.data)))
        .catch((error) => dispatch(actions.getMovieError(error.message)));
};

export const addMovie = (movie) => (dispatch) => {
    movie.genres = movie.genres.map(item => item.value);
    axios.post('http://localhost:4000/movies', movie)
        .then((response) => dispatch(actions.addMovieSuccess(response.data)))
        .catch((error) => dispatch(actions.addMovieError(error.message)));
};

export const updateMovie = (movieId, movie) => (dispatch) => {
    movie.id = movieId;
    movie.genres = movie.genres.map(item => item.value);
    axios.put('http://localhost:4000/movies', movie)
        .then((response) => dispatch(actions.updateMovieSuccess(response.data)))
        .catch((error) => dispatch(actions.updateMovieError(error.message)));
};

export const deleteMovie = (id) => (dispatch) => {
    axios.delete('http://localhost:4000/movies/' + id)
        .then(() => dispatch(actions.deleteMovieSuccess()))
        .catch((error) => dispatch(actions.deleteMovieError(error.message)));
};