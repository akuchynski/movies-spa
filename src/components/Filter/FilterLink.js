import React from "react";
import { useDispatch } from "react-redux";
import Link from 'next/link';
import { getMoviesByParams } from "../../store/thunks";

export const FilterLink = ({ genre }) => {

    const dispatch = useDispatch();

    const isActive = true;

    const handleMenuClick = () => {
        dispatch(getMoviesByParams(searchQuery, genre));
    };

    return (
        <ul>
            <Link
                href={`?genre=${genre}`}
                key={genre}
                class={isActive ? "active" : undefined}
                onClick={handleMenuClick}
            >
                {genre}
            </Link>
        </ul>
    );
};