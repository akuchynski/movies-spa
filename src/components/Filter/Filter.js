import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { useRouter } from 'next/router';
import { getMoviesByParams } from "../../store/thunks";
import { FilterLink } from './FilterLink';
import { genres } from '../../utils/genres';
import { filters } from '../../utils/filters';

export const Filter = () => {

    const dispatch = useDispatch();
    const { query } = useRouter();
    const { name: [name] = [''], genre } = query;

    const [selectedOption, setSelectedOption] = useState(filters[0].value);

    const handleChange = (event) => {
        setSelectedOption(event.target.value);
        dispatch(getMoviesByParams(name, genre, event.target.value));
    };

    return (
        <div className="filter">
            <div className="genre-filter">
                <ul>
                    {genres.map(({ value }) => (<FilterLink key={value} genreValue={value} />))}
                </ul>
            </div>
            <div className="date-filter">
                <span className="sort">SORT BY</span>
                <select value={selectedOption} onChange={handleChange}>
                    {filters.map(({ value, label }) => (
                        <option key={value} value={value}>{label}</option>
                    ))}
                </select>
            </div>
        </div>
    );
};