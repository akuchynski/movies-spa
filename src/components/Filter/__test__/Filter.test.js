import React from "react";
import { render } from "@testing-library/react";
import { Provider } from 'react-redux';
import { Filter } from "../Filter";
import { store } from "../../../store/configureStore";

describe('filter tests', () => {
    it('should render filter component', () => {
        const { asFragment } = render(
            <Provider store={store}>
                <Filter />
            </Provider>);
        expect(asFragment()).toMatchSnapshot();
    });
});