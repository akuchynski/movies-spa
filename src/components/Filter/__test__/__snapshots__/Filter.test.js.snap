// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter tests should render filter component 1`] = `
<DocumentFragment>
  <div
    class="filter"
  >
    <div
      class="genre-filter"
    >
      <ul>
        <li>
          <a
            href="/?genre=Adventure"
          >
            Adventure
          </a>
        </li>
        <li>
          <a
            href="/?genre=Comedy"
          >
            Comedy
          </a>
        </li>
        <li>
          <a
            href="/?genre=Crime"
          >
            Crime
          </a>
        </li>
        <li>
          <a
            href="/?genre=Documentary"
          >
            Documentary
          </a>
        </li>
        <li>
          <a
            href="/?genre=Fantasy"
          >
            Fantasy
          </a>
        </li>
        <li>
          <a
            href="/?genre=Horror"
          >
            Horror
          </a>
        </li>
      </ul>
    </div>
    <div
      class="date-filter"
    >
      <span
        class="sort"
      >
        SORT BY
      </span>
      <select>
        <option
          value="vote_average"
        >
          RATING
        </option>
        <option
          value="release_date"
        >
          DATE
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
